### ISSUE TITLE
New ebean leaves connections with active transactions

### ISSUE THREAD

**Issue by Ryszard-Trojnacki** [ebean-orm/avaje-ebeanorm#232]
I have tested two ebean versions: 3.4.1 and 4.5.3 for transactions.

The code is:
```java
	@Inject
	private TestTableDA tda;

	private static class TransactionInfo {
		public final String id;
		public final String status;
		public final String info;
		
		public TransactionInfo(String id, String status, String info) {
			this.id=id; this.status=status; this.info=info;
		}
		
		@Override
		public String toString() {
			return String.format("Transaction(%s, %s, %s)",id,status,info);
		}
	}
	
	private static class MySQLInfo {
		public List<SqlRow> connections;
		
		public List<TransactionInfo> transactions=new ArrayList<>();
	
		public String status;
		
		public MySQLInfo(List<SqlRow> connections, String status) {
			this.connections=connections;
			this.status=status;
			int p=0;
			for(;;) {
				p=status.indexOf("---TRANSACTION ",p);
				if(p==-1) break;
				p+=15;
				int eol1=status.indexOf("\n",p);
				int eol2=status.indexOf("\n",eol1+2);
				
				transactions.add(new TransactionInfo(status.substring(p,p+6).trim(), status.substring(p+8, eol1).trim() , status.substring(eol1+1, eol2).trim()));
			}
		}
		
		public void log() {
//			log.i("Connections: %s",connections);
			log.i("Transactions: %s",transactions);
//			log.i("Status: %s",status);
		}
	}
	
	private MySQLInfo getMySQLStatus() {
		MySQLInfo res=new MySQLInfo(db.createSqlQuery("show processlist").findList(), db.createSqlQuery("SHOW ENGINE INNODB STATUS;").findUnique().getString("status"));
		res.log();
		return res;
	}
	
	@Test
	public void mySQLTransactionTest() throws InterruptedException {
		Thread.sleep(10000);	// wait for all start (backgroud) operations to finish
		
		if(cfg.getGeneralDbType()!=DbType.MySQL) return;	// check only with mySQL run
		getMySQLStatus();
		tda.getAll();
		getMySQLStatus();
	}
```
TestTableDA.java:
```java
	public List<T> getAll() {
		return ebean.find(TestTable.class).findList();
	}
```
In short it just gets information from mySQL server (Transactions) before select, then it makes test select and after select it gets once again information from mySQL server.

In ebean 3.4.1 I always have (I started this 6 times to be sure):
```
main|INFO |13:47:11|TransactionTests - Transactions: ArrayList#2[Transaction(19C835, not started, MySQL thread id 121, OS thread handle 0x7f53c0401700, query id 69191 localhost 127.0.0.1 test), Transaction(19C834, not started, MySQL thread id 122, OS thread handle 0x7f53c04f6700, query id 69190 localhost 127.0.0.1 test)]
main|INFO |13:47:11|TransactionTests - Transactions: ArrayList#2[Transaction(19C835, not started, MySQL thread id 121, OS thread handle 0x7f53c0401700, query id 69196 localhost 127.0.0.1 test), Transaction(19C836, not started, MySQL thread id 122, OS thread handle 0x7f53c04f6700, query id 69197 localhost 127.0.0.1 test)]

```
Two transactions that are not active before and after select - each time.

With ebean 4.5.3 I have (tested also 6 times):
```
main|INFO |13:51:46|TransactionTests - Transactions: ArrayList#2[Transaction(1A3A9C, not started, MySQL thread id 141, OS thread handle 0x7f53c0463700, query id 104650 localhost 127.0.0.1 test), Transaction(1A34EC, ACTIVE 10 sec, MySQL thread id 142, OS thread handle 0x7f53c0527700, query id 104651 localhost 127.0.0.1 test)]
main|INFO |13:51:46|TransactionTests - Transactions: ArrayList#2[Transaction(1A3A9D, ACTIVE 0 sec, MySQL thread id 141, OS thread handle 0x7f53c0463700, query id 104654 localhost 127.0.0.1 test), Transaction(1A34EC, ACTIVE 10 sec, MySQL thread id 142, OS thread handle 0x7f53c0527700, query id 104653 localhost 127.0.0.1 test)]
```
One transaction is ACTIVE before select (from test starting routine which is done at test start), and after select there are two ACTIVE transactions.

This difference between ebean versions makes for me problem with altering tables, because there cannot be any active transaction when altering:
```
mysql> show processlist;
+----+------+-----------------+------------+---------+------+---------------------------------+-----------------------------------------+
| Id | User | Host            | db         | Command | Time | State                           | Info                                    |
+----+------+-----------------+------------+---------+------+---------------------------------+-----------------------------------------+
| 18 | root | localhost:60603 | documaster | Query   |   33 | Waiting for table metadata lock | ALTER TABLE test_table DROP COLUMN name |
| 19 | root | localhost:60604 | documaster | Sleep   |   97 |                                 | NULL                                    |
| 20 | root | localhost:60605 | NULL       | Query   |    0 | NULL                            | show processlist                        |
+----+------+-----------------+------------+---------+------+---------------------------------+-----------------------------------------+
```
I'm not sure if leaving active transactions doesn't make anything other (slowing database?).

**Comment by rbygrave** [comment]
Try setting a system property:
ebean.transaction.onqueryonly=ROLLBACK

System.setProperty("ebean.transaction.onqueryonly", "ROLLBACK");

... This property defaults to CLOSE in 4.x

**Comment by Ryszard-Trojnacki** [comment]
That worked for me (thanks a lot), but isn't it bad for database performace that transactions are active?

Like you can see in my logs. I have Thread.sleep(10000) command to wait for background processes to finish
```
main|INFO |13:51:46|TransactionTests - Transactions: ArrayList#2[Transaction(1A3A9C, not started, MySQL thread id 141, OS thread handle 0x7f53c0463700, query id 104650 localhost 127.0.0.1 test), Transaction(1A34EC, ACTIVE 10 sec, MySQL thread id 142, OS thread handle 0x7f53c0527700, query id 104651 localhost 127.0.0.1 test)]
```
and it makes transaction active for that whole time "**ACTIVE 10 sec**" and propably much more...

When I was doing alters to database it waited for about 90 seconds (mysql config timeout?).

**Comment by rbygrave** [comment]
If we were running at any isolation level higher than READ_COMMITTED then
we would have to rollback any read only transactions.  However, we
basically have to run at read committed isolation level (otherwise
optimistic concurrency checking does not work) and that is actually defined
in the JPA spec (to run at read committed).

The justification for not rolling back read only transactions when running
at read committed isolation level is that then the "transaction start" is
irrelevant to the execution of statements (well, in Oracle speak we get
'statement level read consistency', the view of the db is at statement
start time and not transaction start time).

Now that said, this obviously effects MySql where it should have no effect
on Oracle or Postgres say. So yes, it might be that for MySql we need to
change the default to ROLLBACK (and for Oracle/Postgres we can leave it as
is) so I'll have a look at that.

**Comment by rbygrave** [comment]
SUMMARY:

The OnQueryOnly.CLOSE optimisation is by default only on for Oracle and Postgres. Can be set on the DatabasePlatform or System property "ebean.transaction.onqueryonly"



DETAIL:

Ok, those 2 commits don't have good 'diffs' as I changed the line endings ... however, the guts of the change is:
- OnQueryOnly moved to DatabasePlatform
- DatabasePlatform defaults to use OnQueryOnly.ROLLBACK
- Oracle and Postgres default to use the OnQueryOnly.CLOSE optimisation

The guts of the OnQueryOnly mode selection is (still) in TransactionManager and is:


  /**
   * Return the behaviour to use when a query only transaction is committed.
   * <p>
   * There is a potential optimisation available when read committed is the default
   * isolation level. If it is, then Connections used only for queries do not require
   * commit or rollback but instead can just be put back into the pool via close().
   * </p>
   * <p>
   * If the Isolation level is higher (say SERIALIZABLE) then Connections used
   * just for queries do need to be committed or rollback after the query.
   * </p>
   */
  private OnQueryOnly initOnQueryOnly(OnQueryOnly dbPlatformOnQueryOnly, DataSource ds) {

    // first check for a system property 'override'
    String systemPropertyValue = System.getProperty("ebean.transaction.onqueryonly");
    if (systemPropertyValue != null) {
      return OnQueryOnly.valueOf(systemPropertyValue.trim().toUpperCase());
    }

    if (OnQueryOnly.CLOSE.equals(dbPlatformOnQueryOnly)) {
      // check for read committed isolation level
      if (!isReadCommitedIsolation(ds)) {
        logger.warn("Ignoring DatabasePlatform.OnQueryOnly.CLOSE as the transaction Isolation Level is not " +
								"READ_COMMITTED");
        // we will just use ROLLBACK and ignore the desired optimisation
        return OnQueryOnly.ROLLBACK;
      } else {
        // will use the OnQueryOnly.CLOSE optimisation
        return OnQueryOnly.CLOSE;
      }
    }
    // default to rollback if not defined on the platform
    return dbPlatformOnQueryOnly == null ? OnQueryOnly.ROLLBACK : dbPlatformOnQueryOnly;
  }

**Comment by rbygrave** [comment]
Closing this as 'fixed' in 4.5.4 (OnQueryOnly.CLOSE optimisation only using for Oracle and Postgres by default).
